File: src/handoff/handoff.js
// handoff.js — feeds/speeds → Creator handoff contract.
//
// The Feeds & Speeds calculator (a separate app, archived for a later phase)
// writes a small advisory payload to localStorage under `gcodeHandoff`. The
// Creator detects it, shows a banner, and can apply the suggested spindle speed
// and feed rate. This module is the DOM-free, storage-agnostic core of that
// behavior so it can be unit-tested without a browser.
//
// Handoff payload shape (all advisory — "suggested starting point"):
//   { rpm: number, feed: number, units: 'mm'|'in', material?: string }

export const HANDOFF_KEY = 'gcodeHandoff';

/**
 * Read and validate a handoff payload from a Storage-like object.
 * Returns the parsed handoff, or null when absent/invalid. Never throws.
 * @param {Storage} storage  e.g. window.localStorage
 * @returns {{rpm:number, feed:number, units:string, material?:string}|null}
 */
export function readHandoff(storage) {
  if (!storage) return null;
  try {
    const raw = storage.getItem(HANDOFF_KEY);
    if (!raw) return null;
    const h = JSON.parse(raw);
    // A valid handoff must at least carry an rpm (mirrors the original guard).
    return h && h.rpm ? h : null;
  } catch (e) {
    return null;
  }
}

/** Remove the handoff payload from storage. Never throws. */
export function clearHandoff(storage) {
  if (!storage) return;
  try {
    storage.removeItem(HANDOFF_KEY);
  } catch (e) {
    /* ignore */
  }
}

/** Human-readable summary for the banner, e.g. "S18000 RPM · F1200 mm/min · MDF". */
export function handoffText(handoff) {
  if (!handoff) return '';
  const rate = handoff.units === 'mm' ? 'mm/min' : 'in/min';
  let text = `S${handoff.rpm} RPM · F${handoff.feed} ${rate}`;
  if (handoff.material) text += ` · ${handoff.material}`;
  return text;
}

/** True when a move type is a cutting move that takes a feed rate. */
function isCutMove(type) {
  return type === 'G1' || type === 'G2' || type === 'G3';
}

/**
 * Compute the state patch produced by applying a handoff. Pure: returns a new
 * partial-state object and does not mutate `state`.
 *
 * Effects (mirrors the original Creator):
 *  - switch to CNC mode
 *  - set spindle RPM and etch feed from the handoff
 *  - set units from the handoff
 *  - fill the feed on any cut move that is currently missing/non-positive
 *
 * @param {Object} state  Current Creator state ({ moves, ... }).
 * @param {Object} handoff  Validated handoff payload.
 * @returns {Object} partial state to merge into the Creator state.
 */
export function applyHandoffToState(state, handoff) {
  const moves = (state.moves || []).map((m) => {
    const missingFeed = m.f === '' || m.f == null || Number(m.f) <= 0;
    if (isCutMove(m.type) && missingFeed) return { ...m, f: String(handoff.feed) };
    return m;
  });
  return {
    machine: 'genericCnc',
    spindleRpm: String(handoff.rpm),
    etchFeed: String(handoff.feed),
    units: handoff.units === 'mm' ? 'mm' : 'in',
    moves,
  };
}

File: python/cam_creation_studio/handoff/handoff.py
"""Feeds/speeds -> Creator handoff contract.

Mirrors the browser prototype's localStorage ``gcodeHandoff`` behavior, but as a
storage-agnostic, pure core. A handoff is an *advisory* payload:

    {"rpm": 18000, "feed": 1200, "units": "mm", "material": "MDF"}

It is a "suggested starting point that requires operator verification" — never a
hard claim.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Mapping, Optional

HANDOFF_KEY = "gcodeHandoff"


@dataclass(frozen=True)
class Handoff:
    rpm: float
    feed: float
    units: str = "mm"
    material: Optional[str] = None

    @staticmethod
    def from_mapping(data: Mapping[str, Any]) -> Optional["Handoff"]:
        """Build a Handoff from a mapping, or None if it lacks a valid rpm."""
        if not data or not data.get("rpm"):
            return None
        return Handoff(
            rpm=data["rpm"],
            feed=data.get("feed", 0),
            units=data.get("units", "mm"),
            material=data.get("material"),
        )


def parse_handoff(raw: Optional[str]) -> Optional[Handoff]:
    """Parse a JSON handoff string. Returns None on absent/invalid input."""
    if not raw:
        return None
    try:
        data = json.loads(raw)
    except (ValueError, TypeError):
        return None
    if not isinstance(data, Mapping):
        return None
    return Handoff.from_mapping(data)


def handoff_text(handoff: Optional[Handoff]) -> str:
    """Human-readable banner summary, e.g. 'S18000 RPM | F1200 mm/min | MDF'."""
    if handoff is None:
        return ""
    rate = "mm/min" if handoff.units == "mm" else "in/min"
    text = f"S{handoff.rpm} RPM | F{handoff.feed} {rate}"
    if handoff.material:
        text += f" | {handoff.material}"
    return text


def _is_cut_move(move_type: str) -> bool:
    return move_type in ("G1", "G2", "G3")


def apply_handoff_to_config(config: Mapping[str, Any], handoff: Handoff) -> dict:
    """Compute the config patch produced by applying a handoff (pure).

    Effects (mirror the original Creator):
      - switch to CNC mode
      - set spindle RPM and etch feed from the handoff
      - set units from the handoff
      - fill the feed on any cut move missing/non-positive
    """
    moves = []
    for m in config.get("moves", []):
        m = dict(m)
        f = m.get("f", "")
        missing = f in ("", None) or _to_float(f) <= 0
        if _is_cut_move(m.get("type", "")) and missing:
            m["f"] = str(handoff.feed)
        moves.append(m)
    return {
        "machine": "genericCnc",
        "spindleRpm": str(handoff.rpm),
        "etchFeed": str(handoff.feed),
        "units": "mm" if handoff.units == "mm" else "in",
        "moves": moves,
    }


def _to_float(value: Any) -> float:
    try:
        return float(value)
    except (ValueError, TypeError):
        return 0.0
